### ISSUE TITLE
Add SSD KingSpec P3-256 to device database

### ISSUE THREAD

**Issue by android116** [smartmontools/smartmontools#249]
Output `smartctl -x /dev/sda` :

`smartctl 7.2 2021-09-14 r5237 [x86_64-linux-5.14.21-150500.55.44-default] (SUSE RPM)
Copyright (C) 2002-20, Bruce Allen, Christian Franke, www.smartmontools.org

\=== START OF INFORMATION SECTION ===
Device Model:     P3-256
Serial Number:    0013155010976
LU WWN Device Id: 5 000000 000000175
Firmware Version: T180910
User Capacity:    256,060,514,304 bytes [256 GB]
Sector Size:      512 bytes logical/physical
Rotation Rate:    Solid State Device
Form Factor:      2.5 inches
TRIM Command:     Available, deterministic, zeroed
Device is:        Not in smartctl database [for details use: -P showall]
ATA Version is:   ACS-2 (minor revision not indicated)
SATA Version is:  SATA 3.2, 6.0 Gb/s (current: 6.0 Gb/s)
Local Time is:    Mon Mar  4 10:16:35 2024 MSK
SMART support is: Available - device has SMART capability.
SMART support is: Enabled
AAM feature is:   Unavailable
APM feature is:   Unavailable
Rd look-ahead is: Enabled
Write cache is:   Enabled
DSN feature is:   Unavailable
ATA Security is:  Disabled, frozen [SEC2]
Wt Cache Reorder: Enabled

=== START OF READ SMART DATA SECTION ===
SMART overall-health self-assessment test result: PASSED

General SMART Values:
Offline data collection status:  (0x02) Offline data collection activity
                                        was completed without error.
                                        Auto Offline Data Collection: Disabled.
Self-test execution status:      (   0) The previous self-test routine completed
                                        without error or no self-test has ever 
                                        been run.
Total time to complete Offline 
data collection:                (   33) seconds.
Offline data collection
capabilities:                    (0x7b) SMART execute Offline immediate.
                                        Auto Offline data collection on/off support.
                                        Suspend Offline collection upon new
                                        command.
                                        Offline surface scan supported.
                                        Self-test supported.
                                        Conveyance Self-test supported.
                                        Selective Self-test supported.
SMART capabilities:            (0x0003) Saves SMART data before entering
                                        power-saving mode.
                                        Supports SMART auto save timer.
Error logging capability:        (0x01) Error logging supported.
                                        General Purpose Logging supported.
Short self-test routine 
recommended polling time:        (   2) minutes.
Extended self-test routine
recommended polling time:        (   2) minutes.
Conveyance self-test routine
recommended polling time:        (   2) minutes.
SCT capabilities:              (0x0031) SCT Status supported.
                                        SCT Feature Control supported.
                                        SCT Data Table supported.

SMART Attributes Data Structure revision number: 16
Vendor Specific SMART Attributes with Thresholds:
ID# ATTRIBUTE_NAME          FLAGS    VALUE WORST THRESH FAIL RAW_VALUE
  9 Power_On_Hours          -O--C-   100   100   000    -    1860
 12 Power_Cycle_Count       -O--C-   100   100   000    -    18
167 Unknown_Attribute       -O---K   100   100   000    -    0
168 Unknown_Attribute       -O--C-   100   100   000    -    0
169 Unknown_Attribute       PO--C-   094   094   010    -    9043991
171 Unknown_Attribute       -O--CK   000   000   000    -    0
172 Unknown_Attribute       -O--CK   000   000   000    -    0
173 Unknown_Attribute       -O--C-   200   200   000    -    8592031754
175 Program_Fail_Count_Chip -O---K   100   100   010    -    0
180 Unused_Rsvd_Blk_Cnt_Tot PO--CK   100   100   000    -    2406
187 Reported_Uncorrect      -O--CK   100   000   000    -    0
192 Power-Off_Retract_Count -O--C-   100   100   000    -    13
194 Temperature_Celsius     -O---K   030   030   000    -    30 (Min/Max 10/40)
206 Unknown_SSD_Attribute   -O--CK   200   200   000    -    2
207 Unknown_SSD_Attribute   -O--CK   200   200   000    -    32
208 Unknown_SSD_Attribute   -O--CK   200   200   000    -    10
209 Unknown_SSD_Attribute   -O--CK   200   200   000    -    7
210 Unknown_Attribute       -O--CK   200   200   000    -    332
211 Unknown_Attribute       -O--CK   200   200   000    -    228
231 Unknown_SSD_Attribute   PO---K   100   100   005    -    0
241 Total_LBAs_Written      -O--CK   100   100   000    -    2431116368
242 Total_LBAs_Read         -O--CK   100   100   000    -    352533067
245 Unknown_Attribute       -O--CK   100   100   000    -    0
                            ||||||_ K auto-keep
                            |||||__ C event count
                            ||||___ R error rate
                            |||____ S speed/performance
                            ||_____ O updated online
                            |______ P prefailure warning

General Purpose Log Directory Version 1
SMART           Log Directory Version 1 [multi-sector log support]
Address    Access  R/W   Size  Description
0x00       GPL,SL  R/O      1  Log Directory
0x01           SL  R/O      1  Summary SMART error log
0x02           SL  R/O     51  Comprehensive SMART error log
0x03       GPL     R/O     64  Ext. Comprehensive SMART error log
0x04       GPL,SL  R/O      8  Device Statistics log
0x06           SL  R/O      1  SMART self-test log
0x07       GPL     R/O      1  Extended self-test log
0x09           SL  R/W      1  Selective self-test log
0x10       GPL     R/O      1  NCQ Command Error log
0x11       GPL     R/O      1  SATA Phy Event Counters log
0x30       GPL,SL  R/O      9  IDENTIFY DEVICE data log
0x80-0x9f  GPL,SL  R/W     16  Host vendor specific log
0xe0       GPL,SL  R/W      1  SCT Command/Status
0xe1       GPL,SL  R/W      1  SCT Data Transfer

SMART Extended Comprehensive Error Log Version: 1 (64 sectors)
No Errors Logged

SMART Extended Self-test Log Version: 1 (1 sectors)
Num  Test_Description    Status                  Remaining  LifeTime(hours)  LBA_of_first_error
# 1  Short offline       Completed without error       00%      1853         -
# 2  Short offline       Completed without error       00%      1829         -
# 3  Extended offline    Completed without error       00%      1827         -
# 4  Short offline       Completed without error       00%      1805         -
# 5  Short offline       Completed without error       00%      1781         -
# 6  Short offline       Completed without error       00%      1758         -
# 7  Short offline       Completed without error       00%      1734         -
# 8  Short offline       Completed without error       00%      1710         -
# 9  Short offline       Completed without error       00%      1686         -
#10  Short offline       Completed without error       00%      1662         -
#11  Short offline       Completed without error       00%      1638         -
#12  Short offline       Completed without error       00%      1614         -
#13  Short offline       Completed without error       00%      1590         -
#14  Short offline       Completed without error       00%      1566         -
#15  Short offline       Completed without error       00%      1542         -
#16  Short offline       Completed without error       00%      1518         -
#17  Short offline       Completed without error       00%      1494         -
#18  Short offline       Completed without error       00%      1470         -
#19  Short offline       Completed without error       00%      1446         -

SMART Selective self-test log data structure revision number 1
 SPAN  MIN_LBA  MAX_LBA  CURRENT_TEST_STATUS
    1        0        0  Not_testing
    2        0        0  Not_testing
    3        0        0  Not_testing
    4        0        0  Not_testing
    5        0        0  Not_testing
Selective self-test flags (0x0):
  After scanning selected spans, do NOT read-scan remainder of disk.
If Selective self-test is pending on power-up, resume after 0 minute delay.

SCT Status Version:                  3
SCT Version (vendor specific):       1 (0x0001)
Device State:                        Active (0)
Current Temperature:                    30 Celsius
Power Cycle Min/Max Temperature:      ?/30 Celsius
Lifetime    Min/Max Temperature:      ?/ ? Celsius
Under/Over Temperature Limit Count:   0/0

SCT Temperature History Version:     2
Temperature Sampling Period:         1 minute
Temperature Logging Interval:        1 minute
Min/Max recommended Temperature:     -127/127 Celsius
Min/Max Temperature Limit:           -127/127 Celsius
Temperature History Size (Index):    478 (125)

Index    Estimated Time   Temperature Celsius
 126    2024-03-04 02:19    30  ***********
 ...    ..(476 skipped).    ..  ***********
 125    2024-03-04 10:16    30  ***********

SCT Error Recovery Control command not supported

Device Statistics (GP Log 0x04)
Page  Offset Size        Value Flags Description
0x01  =====  =               =  ===  == General Statistics (rev 1) ==
0x01  0x008  4              18  ---  Lifetime Power-On Resets
0x01  0x010  4            1860  ---  Power-on Hours
0x01  0x018  6      2431116368  ---  Logical Sectors Written
0x01  0x020  6        47590132  ---  Number of Write Commands
0x01  0x028  6       352533067  ---  Logical Sectors Read
0x01  0x030  6         4880409  ---  Number of Read Commands
0x01  0x038  6   1716093758708  ---  Date and Time TimeStamp
0x07  =====  =               =  ===  == Solid State Device Statistics (rev 1) ==
0x07  0x008  1               4  N--  Percentage Used Endurance Indicator
                                |||_ C monitored condition met
                                ||__ D supports DSN
                                |___ N normalized value

Pending Defects log (GP Log 0x0c) not supported

SATA Phy Event Counters (GP Log 0x11)
ID      Size     Value  Description
0x0001  2            0  Command failed due to ICRC error
0x0003  2            0  R_ERR response for device-to-host data FIS
0x0004  2            0  R_ERR response for host-to-device data FIS
0x0006  2            0  R_ERR response for device-to-host non-data FIS
0x0007  2            0  R_ERR response for host-to-device non-data FIS
0x0008  2            0  Device-to-host non-data FIS retries
0x0009  4            0  Transition from drive PhyRdy to drive PhyNRdy
0x000a  4            3  Device-to-host register FISes sent due to a COMRESET
0x000f  2            0  R_ERR response for host-to-device data FIS, CRC
0x0010  2            0  R_ERR response for host-to-device data FIS, non-CRC
0x0012  2            0  R_ERR response for host-to-device non-data FIS, CRC
0x0013  2            0  R_ERR response for host-to-device non-data FIS, non-CRC
`

**Comment by chrfranke** [comment]
Attribute set is unknown.

**Comment by jonathan-teh** [comment]
This looks like a match for `Lexar 128GB SSD` except that attributes 241 and 242 are supposed to be in GiB for the Lexar (and is being fixed in a pull request) whereas they are in LBA here.
This [article](https://www.redsharknews.com/technology/item/5680-is-it-worth-buying-ultra-cheap-ssds) from July 2018 has an extract of the specifications and notes that the controller can be MAS0902/MK8115/INIC6081, the first two being from Maxiotek. Perhaps this is another (earlier) variant of Maxio attributes?

**Comment by jonathan-teh** [comment]
This attribute set matches the KingSpec `NT-512/T180731` (note the identical first 4 characters of the firmware version) in trac ticket [1174](https://www.smartmontools.org/ticket/1174).

**Comment by jonathan-teh** [comment]
Some background information, there's a [June 2017 review](https://www.thessdreview.com/our-reviews/sata-3/maxiotek-mk8115-controller-preview-mlc-tlc-nand-put-test/2/) of the MK8115 controller with a screenshot of CrystalDiskInfo showing a firmware version of `T170206b` and a similar set of attributes except that it's missing attributes 209-211 which are the SLC erase counts though perhaps they are not relevant for drives with MLC NAND.
Then a [December 2018 post](https://mysku.club/blog/aliexpress/68915.html#comment2943365) on a review of a KingSpec P3 1TB with firmware version `T180731` with `smartctl -x` output that I've posted here: [smartctl-KingSpec-P3-1TB.txt](https://github.com/user-attachments/files/23699917/smartctl-KingSpec-P3-1TB.txt). The review has a teardown that shows a Maxio MAS0902A controller.
Then a [January 2019 post](https://mysku.club/blog/china-stores/69520.html#comment2964386) shows `jm_id` output of a KingSpec P3 128GB with `T180910`:
```
0.011a
Drive: 1(ATA)
Model: P3-128
Fw: T180910
Size: 122104 MB
Firmware id string[2D0]: MKSSD_202000000005400000,Aug 27 2018,09:03:36,MK8215,DCMFCACC
Project id string[280]: v:/MAS090x/Formal/B0KB/For_Customer/Rls_Bin/B0KB_V2.02_SVN0543_MP3
```
confirming a MAS090x controller. So the `T1[0-9]{5}` firmware version is likely date-based.
Now that `Maxio based SSDs` has been merged, I can create a copy of that and have attributes 241 and 242 left at the default Total LBA Written/Read for these `T1[89][1-9]{4}` firmware versions. The hardest part is, what to name it?

1. `Maxio based SSDs (older firmware)`
2. `Maxio MK8115/MAS0902 based SSDs` and the rename the following 2 to `MAS0902/MAS1102` and `MAS1102` but I don't know for sure that they only apply to those specific controllers.
3. `Maxio based SSDs (variant 1)` and rename the following 2 to variants `2` and `3`.

**Comment by jonathan-teh** [comment]
The KingSpec `NT-512/T180731` in trac ticket [1174](https://www.smartmontools.org/ticket/1174) should already match the `JMicron/Maxiotek based SSDs` model family based on:
```
    "[MN]T-(64|128|256|512|[124]TB?)( 22..)?|" // KingSpec MT/NT, MAS0902A-B2C or CS1802A-B2C, tested with
      // MT-512/SN12574, NT-512/T191212, NT-1TB 2242/SN12574
```
as that does not have a firmware version filter but I can't find the `NT-512/T191212` and `NT-1TB 2242/SN12574` in the trac tickets.

**Comment by jonathan-teh** [comment]
I cloned the `linuxhw/SMART` repo (link from another issue) and checked the various KingSpec MT/NT drives there.
The MT/NT entries in `drivedb.h` with firmware `SN12574`  belong in `Maxio based SSDs` as attributes 241, 242 are in GiB.
Can't find one with firmware `T191212`, across any brand, the closest are `T190222` and `T190429` as well as an Apacer AS340 with `T190103` and a Netac with `T190723`, all of which belong in the "older firmware" category proposed here as they have attributes 206-211 which are missing in `JMicron/Maxiotek based SSDs`.
KingSpec MT/NT are also available with Silicon Motion (firmware ending in `A0`) and Realtek (firmware with `0R` from the 3rd character) controllers so they definitely need to move out of `JMicron/Maxiotek based SSDs` (some of the logs with non-Maxio controllers did have unintentional matches).

**Comment by jonathan-teh** [comment]
In light of the `Lexar SSD NS100 256GB/SN17776` in #469 representing yet another attribute variant of Maxio SSDs, I suggest option 3 for naming the model families where the NS100 will become variant 4.
@chrfranke please could you attach the smartctl outputs for `NT-512/T191212` and `NT-1TB 2242/SN12574`, I can then raise a single PR to rename the existing Maxio model families to variants 2 and 3, add variant 1 for this and variant 4 for the NS100, and recategorise the MT/NT entries accordingly.
